### PR TITLE
Fastsim newcfg premix march19

### PR DESCRIPTION
**Pull request by lveldere**
Note: this PR is not ready for review.
It's only here to bookkeep discussions.

This PR will enable premixing for FastSim.
To enable premixing, a serious re-write of the non-simulation cfgs of FastSim was required.

### CODE REVIEW

**Comment by lveldere** [comment]
@Martin-Grunewald  @perrotta 

Hi Andrea and Martin

I am preparing FastSim for premixing with this PR.
This is a pretty urgent matter as we need premixing to start the FastSim production in 74.
Development and initial validation will happen in 75X though.
There are a couple of HLT related things I could use your advice and help.
Could you have a look at the items below?
Some of the items require action. Could you let me know whether you agree with the proposed changes, whether you have perhaps better alternatives, whether and when you would be prepared to implement those changes?

Let me know if clarifications are needed.

Many thanks in advance

Lukas

i) (action required)
muonDigis have changed label: "simMuon{DT,CSC,RPC}Digis" -> "muon{DT,CSC,RPC}Digis"
  would you mind propagating that to  HLTrigger.Configuration.HLT_GRun_Famos_cff? 

ii) (action required)
There is an hlt producer that causes crashes, segfaults,
hltDeDxEstimatorProducer
DeDx estimation in FastSim makes no sense (FastSim has no tracker cluster charge).
Probably this producer is trying to access something that is not defined in FastSim.
In 74 these producers were not in any path, but in 75 they are:
http://cmslxr.fnal.gov/lxr/source/HLTrigger/Configuration/python/HLT_GRun_Famos_cff.py?v=CMSSW_7_5_X_2015-03-12-0200#32450
and next line
Could you remove these trigger paths for FastSim?

iii) (FYI, no actions required)
I notice that HLTGeneralTracks only included signal tracks (so no pu tracks) so I fixed that:

- I replaced the original HLTGeneralTracks with an EDAlias to the mixed track collection.
https://github.com/cms-sw/cmssw/blob/338792e1ba26c91f727f481ded94de95ca88c165/FastSimulation/Tracking/python/HLTIterativeTracking_cff.py#L11
As a consequence, there are no longer multiple identical copies of track collections in the event (nice!)

- the HLT iterative tracking sequences are now empty, since and EDAlias cannot be part of a Sequence:
https://github.com/cms-sw/cmssw/blob/338792e1ba26c91f727f481ded94de95ca88c165/FastSimulation/Tracking/python/HLTIterativeTracking_cff.py#L31

iv) (action required)
The HLT paths make copies of calorimeter RecHit collections.
I got rid of those copies as follows

- In HLTSetup_cff I replaced those copies with empty sequences, see
https://github.com/lveldere/cmssw/blob/fastsim-newcfg-premix-march19/FastSimulation/HighLevelTrigger/python/RecoLocalCalo_cff.py

- Temporarily, I replaced the usual HLT_GRun_Famos file with a a file that resides in FastSimulation/Configuration/python which imports the usual HLT_GRun_Famos,
then removes the calorimeter RecHit producers from the HLT paths, and replaces them with EDAliases to the original collections.
https://github.com/lveldere/cmssw/blob/fastsim-newcfg-premix-march19/FastSimulation/Configuration/python/HLT_GRun_Famos_cff.py

- We need to avoid this file FastSimulation/Configuration/python/HLT_GRun_Famos_cff.py
Could you replace in HLT_GRun_Famos the rechit producers 
[hltEcalRegionalPi0RecHit,hltHfreco,hltHoreco,hltHbhereco,hltEcalRecHit,hltEcalPreshowerRecHit]
with sequences (note the cases)
[HltEcalRegionalPi0RecHit,HltHfreco,HltHoreco,HltHbhereco,HltEcalRecHit,HltEcalPreshowerRecHit]
that would allow me to define inside RecoLocalCalo_cff.py 
- those sequences as empty
- [hltEcalRegionalPi0RecHit,hltHfreco,hltHoreco,hltHbhereco,hltEcalRecHit,hltEcalPreshowerRecHit]
as edaliases
and FastSimulation/Configuration/python/HLT_GRun_Famos_cff.py would become redundant
Alternatively you could remove [hltEcalRegionalPi0RecHit,hltHfreco,hltHoreco,hltHbhereco,hltEcalRecHit,hltEcalPreshowerRecHit]
from the HLT paths, I'm not sure what is easier...

v) (information request)

FastSim does all steps in one job.
Naively one would think that because of this one can replace
DIG,DIGI2RAW,RAW2DIGI with simply DIGI
(perhaps changing some labels where needed)
However, perhaps historically, FastSim does pack and unpack calorimeter digis, See
https://github.com/lveldere/cmssw/blob/fastsim-newcfg-premix-march19/FastSimulation/Configuration/python/Digi_cff.py#L19
Do you know if this is necessary for the HLT?
In other words does the packing and unpacking have an impact on the information used by HLT and its structure?

**Comment by Martin-Grunewald** [comment]
@lveldere 

Hi Lukas,

The action items are all straightforward for us to do in HLT's "confdb.py" where we
have a special section of FastSim customisation. For (iv) we'd actually remove the
modules, which is both simpler and leads to a leaner py file.

Concerning (v), I do not know - except maybe that some lepton isolation modules
need the detailed calo information - but then only unpacking is needed, not packing,
unless the content/format is changed ... ?? Maybe @perrotta or @giamman, who did
a lot of fastsim in the past, remember?

Best regards

Martin

**Comment by giamman** [comment]
About point (v), this was probably part of the modifications by Federica Primavera in the context of the transition towards the usage of the FullSim digitizers in the calorimeters.
I am not sure, but I suspect that the reason was that some transient product produced by one of those modules was needed by the next steps. Therefore, I suppose that before being able to simplify the workflow one should review the digitizer modules themselves.

**Comment by lveldere** [comment]
Thanks for your replies

@Martin-Grunewald 
Great to hear that this is a straightforward operation.
Actually there are a few more modifications of the same order I would like to propose later today.

**Comment by lveldere** [comment]
@Martin-Grunewald @perrotta 

Hi Martin

First of all a correction with respect to iv):
please do *not* remove 
hltEcalPreshowerRecHit, hltEcalRecHit
I realise now that the trigger menu overwrites the definitions given inside FastSimulation.HighLevelTrigger.HLTSetup_cff
(I'm not sure whether or not that is appropriate, but I would keep that for a future discussion)

I managed to run Reconstruction, L1Reco and HLT without packing / unpacking any digis.
(ECAL and HCAL dpgs confirm this should work fine)
It requires a few extra edits to the HLT menu.
If you allow me, I would like to add 2 additional action items (see below).

I also add one more, which would be to remove a number of empty tracking sequences.

I believe/hope this is all I would need for now.

Many thanks in advance

Lukas

vi)
the hlt ecal RecHit producer needs a few different settings
```
hltEcalRecHit.killDeadChannels = False
hltEcalRecHit.recoverEBFE = False
hltEcalRecHit.recoverEEFE = False
```
this prevents the rechit producer to read in collections of flags for dead, noisy, ... cells produced during unpacking. ECAL dpg confirms that the impact on simulation should be minor.

vii)
the following modules would need to be removed from all trigger paths
hltEcalPreshowerDigis,hltEcalDigis,hltEcalDetIdToBeRecovered
the first two are copies of modules that reside earlier in the schedule.
I will add EDAliases to replace these modules.

viii)
Could you remove the following empty sequences from the HLT paths:
HLTIterativeTrackingIter04
HLTIterativeTrackingIter02
HLTIterativeTracking
HLTIterativeTrackingTau3Mu
HLTIterativeTrackingReg
HLTIterativeTrackingForElectronIter02
HLTIterativeTrackingForPhotonsIter02
HLTIterativeTrackingL3MuonIter02
HLTIterativeTrackingL3MuonRegIter02
HLTIterativeTrackingForBTagIter02
HLTIterativeTrackingForTauIter02
HLTIterativeTrackingForTauIter04
HLTIterativeTrackingGlbTrkMuonIter02
HLTIterativeTrackingHighPtTkMu
HLTIterativeTrackingHighPtTkMuIsoIter02
HLTIterativeTrackingDisplacedJpsiIter02
HLTIterativeTrackingDisplacedPsiPrimeIter02
HLTIterativeTrackingDisplacedNRMuMuIter02
HLTIterativeTrackingForBTagIter12
HLTIterativeTrackingForBTagIteration0
HLTIterativeTrackingIteration4DisplacedJets

**Comment by perrotta** [comment]
Hi Lukas.

As Martin said, the customization for the HLT in FastSim is done with  
the help of the
HLTrigger/Configuration/python/Tools/confdb.py
script, and most of the customised modules are defined within
FastSimulation/HighLevelTrigger/python
configs (or other fastsin configs linked from there).

Said that, everything can be done from that script, the easiest things  
being removing a whole path.

About your last point raised here:

vi) This can also be easily be managed by confdb.py, where the fastsim  
customization coulbe defined inside the FastSimulation area (e.g, now  
it is customized in [0], but you can perhaps find a more convenient  
customization hlt-oriented if you think so)

vii) A hlt path can't change as a whole. What we do is to redefine  
modules and sequences therein in order to be fastsim compliant. For  
example, the modules you want to be "removed" here are all dummified by
FastSimulation/HighLevelTrigger/python/HLTSetup_cff.py
thus there is no need to physically remove them from the path. Other  
modules can be added there, and I personally take care of it every  
time I customize a new HLT path for fastsim, if needed.

viii) I don't know what you plan to do, perhaps it would have been  
better to sit down and discuss the whole strategy all together.  
However, those sequences are not "empty". Take for example  
HLTIterativeTrackingIter02, the main iterative tracking sequence for  
HLT. In the fullsim/online workflow that sequence cooks all the  
necessary ingredients and ends up with a hltIter2Merged track, which  
is then feed up to the subsequent steps in the trigger paths. In  
FastSim the same sequence gives out a hltIter2Merged [1], which is a  
clone of HLTgeneralTracks (and as I told you already, it could even be  
better adapted to the different kind of iterative tracking sequences  
now in the menu case by case, while for the moment an identical clone  
of the same HLTgeneralTracks is taken in all cases [3]).
Therefore, it is not clear to me what do you mean by "Could you remove  
the following empty sequences from the HLT paths"...

Cheers,
Andrea


[0]
http://cmslxr.fnal.gov/source/FastSimulation/HighLevelTrigger/python/HLTFastReco_cff.py#0044

[1]
http://cmslxr.fnal.gov/source/FastSimulation/Tracking/python/HLTIterativeTracking_cff.py#0035

[2]
FastSimulation/Tracking/python/HLTGeneralTracks_cfi.py

[3]
http://cmslxr.fnal.gov/source/FastSimulation/Tracking/python/HLTIterativeTracking_cff.py#0015

lveldere <notifications@github.com> ha scritto:

**Comment by lveldere** [comment]
Hi Andrea, Martin

I second Andrea, it would be excellent if you and I, and perhaps Martin could sit together for a short while. There are a few things in our conversation that confuse me.

Perhaps as a preparation for this meeting:
I think all your concerns and my request can be captured in one example.

In the release, the HLT for FastSim is configured in [1].
In my pr it is configured in [2], which imports [1] and makes some alterations.
In [2], I replace hltEcalDigis, originally defined here [1, line 1462] with an EDAlias to the ecal digis produced by the module with label simEcalDigis in the DIGI step [2, line 73].
(this saves CPU and memory and removes then need for an unnecessary layer of complexity, namely packing and unpacking of digis). I do not see how I can do the same efficient thing with the help of confdb.py or dummies.

BTW: Andrea, the tracking sequences are empty in my pr, because the clones are replaced with EDAliases, which saves memory and CPU, but cannot be inside a sequence.

Cheers

Lukas

[1]
http://cmslxr.fnal.gov/lxr/source/HLTrigger/Configuration/python/HLT_GRun_Famos_cff.py#1462
That means, whatever is done in HLTSetup, 

[2]
https://github.com/lveldere/cmssw/blob/fastsim-newcfg-premix-march19/FastSimulation/Configuration/python/HLT_GRun_Famos_cff.py#L73

**Comment by perrotta** [comment]
@lveldere 

I must admit that I don't know exactly what an EDAlias is, and what does prevent a configuration as the one in [1] of yours from fitting inside FastSimulation/HighLevelTrigger/python/HLTSetup_cff.py , or in a separate cff file imported in it.

However, you could obtain exactly what you have in your new
FastSimulation/Configuration/python/HLT_GRun_Famos_cff.py
by adding the extra lines in it within a 
    if self.config.fastsim:
block inside HLTrigger/Configuration/python/Tools/confdb.py, so that they will be dumped into the HLTrigger/Configuration/python/HLT_GRun_Famos_cff.py when we use that script to automatically create the HLT_GRun_Famos_cff.py configuration.

Ok, I'll try to be more propositive as soon as I understand better what you plan to do (and also what an EDAlias is...).

**Comment by lveldere** [comment]
superseeded by #8455

**Comment by lveldere** [comment]
please test
